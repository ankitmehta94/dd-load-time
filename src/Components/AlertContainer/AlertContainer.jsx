import { useEffect, useState } from "react";
import { checkIfAlertConditionIsTrue, createAlertObject, getNextAlertType } from "../../Utils/AlertingUtils";
import CardHOC from "../Card/Card";
import style from "./AlertContainer.css";
import {
  BELOW_CONDITION_KEY,
  ABOVE_CONDITION_KEY,
  THRESHOLD_DEFAULT,
  SAMPLING_RATE,
  ALERT_WINDOW_DEFAULT,
} from "../../Constants/Constants";
import AlertTable from "../AlertTable/AlertTable";

export const AlertContainerDefaultProps = {
chartData: [],
  className : '',
  threshold : THRESHOLD_DEFAULT,
  timeWindow: ALERT_WINDOW_DEFAULT,
  samplingRate: SAMPLING_RATE,
}
const AlertTableCard = CardHOC(AlertTable);
export default function AlertContainer({
  chartData,
  className,
  threshold,
  timeWindow,
  samplingRate,
}) {
  const [alertState, setalertState] = useState([]);
  useEffect(() => {
    const newAlert = getNextAlertType(alertState)
    const alertConditionMet = checkIfAlertConditionIsTrue(
      chartData,
      newAlert,
      threshold,
      timeWindow / samplingRate
    );
    console.log(alertConditionMet, "<-----------------alertConditionMet", alert);
    if (alertConditionMet) {
      var notificationText =  `Load Time Went ${newAlert} the threshold of ${threshold}`
      var notification = new Notification("Load Time Alert", { body: notificationText });
      alertState.push(createAlertObject(newAlert, threshold, timeWindow));
      setalertState(alertState.slice());
    }
  }, [chartData, threshold, timeWindow, samplingRate, alertState]); // MAybe this isn't correct
  useEffect(() => {
    if(typeof Notification !== 'undefined'){
      Notification.requestPermission().then(function (result) {
        console.log(result);
      });
    }

  }, []);
  const {alert:lastAlert = '' } = alertState[alertState.length-1] || '';
  const [aboveArray, belowArray] = splitAlertArray(alertState)
  return (
    <div className={className}>
      <div className={style["list-container"]}>
        <AlertTableCard
        lastAlert={lastAlert === ABOVE_CONDITION_KEY}
          alertData={aboveArray}
          alertType={ABOVE_CONDITION_KEY}
          infoText={'This shows the list of alerts made in the past and which alerts are currently on going'}
        />
      </div>
      <div className={style["list-container"]}>
        <AlertTableCard
         lastAlert={BELOW_CONDITION_KEY === lastAlert}
          alertData={belowArray}
          alertType={BELOW_CONDITION_KEY}
          infoText={'This shows the list of alerts made in the past and which alerts are currently on going'}
        />
      </div>
    </div>
  );
}

function splitAlertArray(alertArray) {
    const [aboveArray, belowArray] =                            
    alertArray.reduce((result, element) => {
      result[element.alert === ABOVE_CONDITION_KEY ? 0 : 1].push(element);
      return result;
    },
    [[], []]);  
    return [aboveArray, belowArray];
}