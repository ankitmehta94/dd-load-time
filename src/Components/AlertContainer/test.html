<ul class="_7yu" id="js_dz" role="menu"><li class="" style=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">E-3</div></div></li><li class="" style=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">E-3D</div></div></li><li class="" style=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">EAD</div></div></li><li class="" style=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">F-1</div></div></li><li class="" style=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">F-1 CPT</div></div></li><li class="" style=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">F-1 OPT</div></div></li><li class="" style=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">H-1B</div></div></li><li class="" style=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">H-1B1</div></div></li><li class="" style=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">J-1</div></div></li><li class="" style=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">J-2</div></div></li><li class=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">L-1</div></div></li><li class=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">L-2</div></div></li><li class=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">O-1</div></div></li><li class=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">TN</div></div></li><li class=""><div aria-checked="false" aria-disabled="false" class="_2wpb _3v8w" role="menuitemcheckbox" tabindex="-1" style="letter-spacing: normal; font-family: -apple-system, system-ui, BlinkMacSystemFont, Arial, sans-serif; color: rgb(28, 30, 33); font-size: 12px; font-weight: normal; line-height: 16px; padding: 6px 24px 6px 32px;"><div class="_3leq">Other</div></div></li></ul>