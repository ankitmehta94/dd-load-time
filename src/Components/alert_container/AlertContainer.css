.list-container{
    width: 49%;
    height: 98%;
}
.alert{
    margin: 0.5% 0 ;
    padding: 1%;
    width: 97%;
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
    align-items: center;
    border: 1px solid black;
    border-radius: 5px;
}
.alert-table{
    display: flex;
    flex-direction: column;
    width: 100%;
    justify-content: start;
    align-items: center;
}
.alert-pulse{
    animation: pulse 2s infinite;
}
.alert-table-list>:first-child{
    margin-top: 10px;
}
.alert-table-list{
    display: flex;
    flex-direction: column;
    justify-content: start;
    align-items: center;
    width:95%;
    height: 40vh;
    overflow-y: scroll;
    
}
.alert-table-header{
    margin: 0.5% 0 ;
    padding: 1%;
    width: 97%;
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
    align-items: center;
    
}
.red-alert{
    background-color: #f7cfcf;
}
.green-alert{
    background-color: #dcf5dc;
}   
@-webkit-keyframes pulse {
    0% {
      -webkit-box-shadow: 0 0 0 0 rgba(204, 169, 44, 0.4);
    }
    70% {
        -webkit-box-shadow: 0 0 0 10px rgba(204,169,44, 0);
    }
    100% {
        -webkit-box-shadow: 0 0 0 0 rgba(204,169,44, 0);
    }
  }
  @keyframes pulse {
    0% {
      -moz-box-shadow: 0 0 0 0 rgba(204,169,44, 0.4);
      box-shadow: 0 0 0 0 rgba(204,169,44, 0.4);
    }
    70% {
        -moz-box-shadow: 0 0 0 10px rgba(204,169,44, 0);
        box-shadow: 0 0 0 10px rgba(204,169,44, 0);
    }
    100% {
        -moz-box-shadow: 0 0 0 0 rgba(204,169,44, 0);
        box-shadow: 0 0 0 0 rgba(204,169,44, 0);
    }
  }