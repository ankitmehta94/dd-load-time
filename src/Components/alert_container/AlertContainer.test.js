import AlertContainer , {AlertContainerDefaultProps} from './AlertContainer';
import renderer from 'react-test-renderer';

it('Default State of AlertContainer', () => {   
    const DateInputComponent = renderer.create(<AlertContainer {...AlertContainerDefaultProps} />).toJSON();
    expect(DateInputComponent).toMatchSnapshot();
});