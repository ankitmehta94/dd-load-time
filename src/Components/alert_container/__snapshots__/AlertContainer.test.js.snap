// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default State of AlertContainer 1`] = `
<div
  className=""
>
  <div>
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      style={
        Object {
          "height": "100%",
        }
      }
    >
      <div
        className="MuiTableContainer-root makeStyles-container-2"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-flex-end MuiGrid-grid-xs-12"
        >
          <div
            aria-describedby={null}
            className=""
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            title="This shows the list of alerts made in the past and which alerts are currently on going"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-info-circle fa-w-16 fa-lg "
              data-icon="info-circle"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </div>
        </div>
        <table
          aria-label="alert table"
          className="MuiTable-root MuiTable-stickyHeader"
          role={null}
        >
          <thead
            className="MuiTableHead-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head"
              role={null}
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
                scope="col"
              >
                Status
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-stickyHeader"
                scope="col"
              >
                Type
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-stickyHeader"
                scope="col"
              >
                Time
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-stickyHeader"
                scope="col"
              >
                Threshold
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-stickyHeader"
                scope="col"
              >
                Alert Window
              </th>
            </tr>
          </thead>
          <tbody
            className="MuiTableBody-root"
            role={null}
          />
        </table>
      </div>
    </div>
  </div>
  <div>
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      style={
        Object {
          "height": "100%",
        }
      }
    >
      <div
        className="MuiTableContainer-root makeStyles-container-2"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-flex-end MuiGrid-grid-xs-12"
        >
          <div
            aria-describedby={null}
            className=""
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            title="This shows the list of alerts made in the past and which alerts are currently on going"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-info-circle fa-w-16 fa-lg "
              data-icon="info-circle"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </div>
        </div>
        <table
          aria-label="alert table"
          className="MuiTable-root MuiTable-stickyHeader"
          role={null}
        >
          <thead
            className="MuiTableHead-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head"
              role={null}
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
                scope="col"
              >
                Status
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-stickyHeader"
                scope="col"
              >
                Type
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-stickyHeader"
                scope="col"
              >
                Time
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-stickyHeader"
                scope="col"
              >
                Threshold
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-stickyHeader"
                scope="col"
              >
                Alert Window
              </th>
            </tr>
          </thead>
          <tbody
            className="MuiTableBody-root"
            role={null}
          />
        </table>
      </div>
    </div>
  </div>
</div>
`;
